### PR TITLE
planner: make sure that join key type are always the same for mpp join

### PR DESCRIPTION
**Pull request by windtalker**
### What problem does this PR solve?

Issue Number: close pingcap/tics#1733 

Problem Summary:

As the issue described.
### What is changed and how it works?

Proposal: [xxx](url) 

What's Changed:

How it Works:

Add cast for join key to make sure the join key always has the same type for mpp join.
### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test
- Manual test (add detailed scripts or steps below)

Side effects

- Performance regression
    - Consumes more CPU
    - Consumes more MEM
- Breaking backward compatibility

### Release note 

- fix bug that mpp shuffle join may return wrong results if the join key of right/left table does not have exactly the same type.

### CODE REVIEW

**Review by hanfei1991 — APPROVED** [review]

File: planner/core/task.go [hunk 1]
```diff
@@ -561,7 +561,13 @@ func (p *PhysicalHashJoin) attach2Task(tasks ...task) task {
 
 // for different decimal scale and precision, tiflash uses different underlying type.
 // Here we check the scale and precision to decide whether conversion is a must.
```
Comment (fzhedu) [inline planner/core/task.go:563]: change comments as needed pls.

File: planner/core/optimizer_test.go [hunk 2]
```diff
@@ -59,3 +59,53 @@ func (t *testPlannerFunctionSuite) TestMPPDecimalConvert(c *C) {
...
+	unsignedBigIntType.Flen, tinyIntType.Decimal = mysql.GetDefaultFieldLengthAndDecimal(mysql.TypeLonglong)
+	unsignedBigIntType.Flag = mysql.UnsignedFlag
+
+	decimalType := &types.FieldType{
+		Tp:      mysql.TypeNewDecimal,
+		Flen:    20,
+		Decimal: 0,
+	}
+
+	testJoinKeyTypeConvert(tinyIntType, tinyIntType, tinyIntType, false, false, c)
```
Comment (fzhedu) [inline planner/core/optimizer_test.go:102]: what about int/small int/big int vs decimal?

File: planner/core/task.go [hunk 3]
```diff
@@ -581,22 +587,31 @@ func needDecimalConvert(tp *types.FieldType, rtp *types.FieldType) bool {
...
+		cLen = mathutil.Min(65, cLen)
+		commonType.Decimal = cDec
+		commonType.Flen = cLen
+	} else if needConvert(lType, commonType) || needConvert(rType, commonType) {
+		if mysql.IsIntegerType(commonType.Tp) {
+			// If the target type is int, both TiFlash and Mysql only support cast to Int64
+			// so we need to promote the type to Int64
+			commonType.Tp = mysql.TypeLonglong
+			commonType.Flen = mysql.MaxIntWidth
+		}
```
Comment (fzhedu) [inline planner/core/task.go:612]: what if int(x) = decimal(n,m)?

File: planner/core/task.go (hunk 1 above)
Comment (windtalker) [reply planner/core/task.go]: Done.

File: planner/core/optimizer_test.go (hunk 2 above)
Comment (windtalker) [reply planner/core/optimizer_test.go:102]: TiDB guarantees that two type are in the same catalog, so do not need to test integer vs decimal

File: planner/core/task.go (hunk 3 above)
Comment (windtalker) [reply planner/core/task.go:613]: TiDB guarantees that two type are in the same catalog, so it should not happen here.

**Pull request merged**